### PR TITLE
Fix/indexes

### PR DESCRIPTION
**Pull request by mixmix**
This PR is aimed at resolving an index problem we've found.

I've written a test suite which can be used on `ssb-query` and `ssb-backlinks`. Apologies for the complexity, I wrote it so that we can test the same things in multiple times with multiple different states in a way which is easier to read / write

Current state after first 2 commits : have isolated a state which breaks ... and how you can unbreak it (publish more messages, then restart :cry: ) but not the source of the problem

### CODE REVIEW

File: test/indexes-check.test.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+  /* state */
+  var published = [] // contents of messages we published
+  var lastMsgId
+
+  server.tribes.create({}, (err, { groupId } = {}) => {
+    t.error(err, 'creates group')
+
+    var recps = [groupId]
+    // recps = null // passes all tests
```
Comment (mixmix) [inline test/indexes-check.test.js:24]: NOTE: uncomment this line and all the tests pass ):

**Comment by mixmix** [comment]
For sanity, here's the output I see, so we know we're saying the same failures!

```

SSBC/ssb-tribes ‹fix/indexes*› » node test/indexes-check.test.js | tap-spec                              1 ↵

  backlinks

    ✔ creates group
    ✔ encrypted (0, profile/person)
    ✔ encrypted (1, link/feed-profile)
    ✔ encrypted (2, profile/person)
    ✔ encrypted (3, link/feed-profile)
    ✔ all published

  > check backlinks results

    ✔ backlinks.read
    ✔ backlinks finds (1, link/feed-profile)
    ✔ backlinks finds (3, link/feed-profile)

  > check again (after server restart)

    ✔ server restart
    ✔ backlinks.read
    ✔ backlinks finds (1, link/feed-profile)

    ✖ backlinks finds (3, link/feed-profile)
    -----------------------------------------
      operator: deepEqual
      expected: |-
        { type: 'link/feed-profile', count: 3, parent: '@wafO5qrMr2zhwYhMSPfswG+qyEYKq/jGn7+UQSVgNkI=.ed25519', child: '%joULlZB9IiOkJIYn4W2huQb5ocJh7KuDQhwM8B7ly9Y=.sha256', recps: [ '%C8CcxG9I79gw1azVI5df5IZR/sfELwCrzbbYft9tE7s=.cloaked' ], tangles: { group: { root: '%+LTPA1saMF50kV7+bki9hcaCsaNDsUzqGwDeVdLxfBY=.sha256', previous: [ '%joULlZB9IiOkJIYn4W2huQb5ocJh7KuDQhwM8B7ly9Y=.sha256' ] } } }
      actual: |-
        undefined
      at: _published.forEach (/home/mix/projects/SSBC/ssb-tribes/test/indexes-check.test.js:68:13)
      stack: |-
  Error: backlinks finds (3, link/feed-profile)
  at Test.assert [as _assert] (/home/mix/projects/SSBC/ssb-tribes/node_modules/tape/lib/test.js:228:54)
  at Test.bound [as _assert] (/home/mix/projects/SSBC/ssb-tribes/node_modules/tape/lib/test.js:80:32)
  at Test.tapeDeepEqual (/home/mix/projects/SSBC/ssb-tribes/node_modules/tape/lib/test.js:426:10)
  at Test.bound [as deepEqual] (/home/mix/projects/SSBC/ssb-tribes/node_modules/tape/lib/test.js:80:32)
  at _published.forEach (/home/mix/projects/SSBC/ssb-tribes/test/indexes-check.test.js:68:13)
  at Array.forEach (<anonymous>)
  at pull.collect (/home/mix/projects/SSBC/ssb-tribes/test/indexes-check.test.js:67:20)
  at /home/mix/projects/SSBC/ssb-tribes/node_modules/pull-stream/sinks/reduce.js:10:5
  at /home/mix/projects/SSBC/ssb-tribes/node_modules/pull-stream/sinks/drain.js:20:24
  at /home/mix/projects/SSBC/ssb-tribes/node_modules/pull-stream/throughs/through.js:19:9


  > check AGAIN (after a publish + restart)

    ✔ publish anything
    ✔ server restart
    ✔ backlinks.read
    ✔ backlinks finds (1, link/feed-profile)
    ✔ backlinks finds (3, link/feed-profile)



  Failed Tests: There was 1 failure

    > check again (after server restart)

      ✖ backlinks finds (3, link/feed-profile)
```

See the same this with the "query" test

File: test/indexes-check.test.js (hunk 1 above)
Comment (mixmix) [reply test/indexes-check.test.js:24]: Also just checked with `ssb-private1` style encryption, same problem there.
So it's something to do with the box/ unbox pipeline ...

**Comment by mixmix** [comment]
**ideas of things to test/ investigate**:
- what is actually in the index at different times?
    - I tried looking at `sbot.status().sync` but it's not that useful ... unless it's put deep in the `backlinks.read`?
    - other than looking at `since` can we check how `level` is writing and calling back at a low level? (in `flumeview-level`)?
- what does restarting the server do?
    - if you publish an extra "anything" message and then query again, that still fails
    - something about being not the last message, and the restart moves something along?
- what's in the log?
    - we assume the message makes it into the log, because on restart it's indexed sometimes

This is a list of questions I came up with at the end of the day. I wish I'd made a list like this earlier to systematically work through. I recommend you edit + add questions to this before you get lost in it!

**Comment by christianbundy** [comment]
Resolution: https://github.com/ssbc/ssb-db/pull/306

**Pull request merged**